internal/repository/repository.go

```diff
@@ -515,6 +515,92 @@ func (r *Repository) LoadIndex(ctx context.Context) error {
 	return nil
 }
 
+// LoadIndexFromPacks loads a new index by reading all pack files
+// This will create one Index data structure per pack file readed that is not
+// marked as finalized
+func (r *Repository) LoadIndexFromPacks(ctx context.Context, packs restic.IDSet, ids restic.IDs, p *restic.Progress) error {
+	p.Start()
+	defer p.Done()
+
+	debug.Log("Loading index from pack files")
+
+	// track spawned goroutines using wg, create a new context which is
+	// cancelled as soon as an error occurs.
+	wg, ctx := errgroup.WithContext(ctx)
+
+	type FileInfo struct {
+		restic.ID
+		Size int64
+	}
+	ch := make(chan FileInfo)
+	indexCh := make(chan *Index)
+
+	// send list of index files through ch, which is closed afterwards
+	wg.Go(func() error {
+		defer close(ch)
+		return r.List(ctx, restic.PackFile, func(id restic.ID, size int64) error {
+			if !packs.Has(id) {
+				return nil
+			}
+			select {
+			case <-ctx.Done():
+				return nil
+			case ch <- FileInfo{id, size}:
+			}
+			return nil
+		})
+	})
+
+	// a worker receives an index ID from ch, reads the pack contents, and sends it to indexCh
+	worker := func() error {
+		for fi := range ch {
+			entries, _, err := r.ListPack(ctx, fi.ID, fi.Size)
+			idx := NewIndex()
+			idx.StorePack(fi.ID, entries)
+
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("unable to load index %v", fi.ID.Str()))
+			}
+
+			select {
+			case indexCh <- idx:
+			case <-ctx.Done():
+			}
+		}
+
+		return nil
+	}
+
+	// run workers on ch
+	wg.Go(func() error {
+		defer close(indexCh)
+		return RunWorkers(loadIndexParallelism, worker)
+	})
+
+	// receive decoded indexes
+	wg.Go(func() error {
+		for idx := range indexCh {
+			r.idx.Insert(idx)
+			p.Report(restic.Stat{Blobs: 1})
+		}
+		// Add new finalized index to set given index IDs
+		for _, id := range ids {
```

Comment:
I don't like this hack to inject the index files list into the index. Wouldn't it be cleaner to pass the list of obsolete index files directly to `rebuildIndexFiles` / `MasterIndex.Save`? Or add a new constructor to `Index` which allows directly creating a finalized index with a list of superseded packs?

Reply:
You are right, this is a bad hack :wink: 
I added `extraObsolete` to `MasterIndex.Save` to solve this.